<a href="https://colab.research.google.com/github/Luisadle/Business-Intelligence/blob/main/Reto_2_Luisa_De_La_Espriella.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Reto #2 - Árboles de Decisiones**
0. Se procede con lacarga de las librerias de trabajo

In [ ]:
#Librerias cientificas
import numpy as np
import pandas as pd

#liberias Arbol
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

#Conectar al Drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


1. Se procede a cargar los datos

In [ ]:
nxl='/content/drive/MyDrive/Colab Notebooks/2. CollegeAdmisions_Int_M.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)
XDB.head(100)
XDB=XDB.dropna()

#Seleccionar las variables
XD=XDB[['Edu_Parent1','Edu_Parent2','Gender','White','Asian','HSGPA','SAT/ACT']]
yd=XDB['Admitted']
XD=XD.replace({'M': 1, 'F': 0}) #Se usa para convertir las variables M y F, en variables numéricas

<ipython-input-4-198dffde209c>:9: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  XD=XD.replace({'M': 1, 'F': 0}) #Se usa para convertir las variables M y F, en variables numéricas


In [ ]:
totaldatos=len(XDB) #Cantidad de datos en la base de datos
print("El total de datos es:\n", totaldatos)

cantidadV = yd.value_counts() #Frecuencia de la variable
print("La cantidad de datos en la variable Admitted es:\n", cantidadV)

pv = cantidadV/len(yd) *100 #Probabilidad de la variable
print("La probabilidad de la variable es:\n", pv)

El total de datos es:
 17339
La cantidad de datos en la variable Admitted es:
 Admitted
No     12016
Yes     5323
Name: count, dtype: int64
La probabilidad de la variable es:
 Admitted
No     69.300421
Yes    30.699579
Name: count, dtype: float64


**Descripción de los datos:** De acuerdo con la base de datos de Admisiones a la Universidad, esta posee un total de 17339 datos totales, divididos en 12016 de solicitantes que no fueron admitidos y 5323 de solicitades que si fueron admitidos. Cada uno de los solicitantes de admisión estará descrito en función de 8 variables:


*   Edu_Parent 1: Esta establece el grado de educación de los padres de secundaria y media. Es clasificado de la siguiente manera:
  *   1 - Sin escuela preparatoria
  *   2 - Escula preparatoria trunca
  *   3 - Graduado de la escula preparatoria

*   Edu_Parent 2: Esta establece el grado de educación de los padres en la educación superior. Es clasificado de la siguiente manera:
  *   4 - Universidad trunca
  *   6 - Graduado universitario en carrera de 4 años
  *   7 - Posgrado

*   Gender: Reconoce el genero del aplicante, ya sea Femenino o Masculino.

*   White: Reconoce si el aplicante es blanco o no.

*   Asian: Reconoce si el aplicante es asiatico o no.

*   HSGPA: Es el GPA ponderado de la escula preparatoria del aplicante. Va de 0 a 5.

*   SAT/ACT: Es la puntuación más alta de SAT/ACT, donde la puntuación de ACT se convierte primero en la puntuación del SAT equivalente para inglés y matemáticas.

*   Admitted: Establece si el aplicante fue admitido o no a la carrera establecida por el mismo.






2. Se procede con la implementación del Modelo

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
mar.fit(XD,yd)

#Se procede con la gráfica
from pydotplus import graph_from_dot_data #Me crea la gráfica
from sklearn.tree import export_graphviz #Me exporta la gráfica aPNG, JPEG, TIFF

ve=['Edu_Parent1','Edu_Parent2','Gender','White','Asian','HSGPA','SAT/ACT'] #Nombre para el árbol
dot_data=export_graphviz(mar,feature_names=ve)#crear gráfico
graph=graph_from_dot_data(dot_data)
graph.write_png('ÁrbolDecisión - Reto2.0.png') #Crea una imagen del árbol
#La imagen del árbol de decisiones se encuentra en la carpeta CONTENT


True

3. Se procede con el pronóstico de los aplicantes

In [ ]:
XDB0=pd.read_excel(nxl,sheet_name=1)
XDB0.head(100)

#Se crea la base de datos a pronosticar
XD0=XDB0[['Edu_Parent1','Edu_Parent2','Gender','White','Asian','HSGPA','SAT/ACT']]
# Replace 'M' and 'F' in 'Gender' column with 1 and 0 respectively
XD0 = XD0.replace({'Gender': {'M': 1, 'F': 0}})
yd0=XDB0['Admitted']


#Se Calcula la admisión a los aplicantes
y_pred=mar.predict(XD0)
print(y_pred)

#Se Crea el informe en excel
df=pd.DataFrame(np.column_stack((XD0, y_pred)))
df.columns=['Edu_Parent1','Edu_Parent2','Gender','White','Asian','HSGPA','SAT/ACT', 'Admitted']
df.to_excel('Admisiones de Universidades.xlsx')

['No' 'Yes' 'No' 'No' 'No' 'No' 'No' 'No' 'No' 'No' 'Yes']


<ipython-input-41-feec2975c498>:7: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  XD0 = XD0.replace({'Gender': {'M': 1, 'F': 0}})


In [ ]:
totaldatos1=len(XDB0) #Cantidad de datos en la base de datos
print("El total de datos es:\n", totaldatos1)

cantidadN = pd.Series(y_pred).value_counts() #Frecuencia de la variable
print("La cantidad de datos en la variable Admitted es:\n", cantidadN)

pvN = cantidadN/len(y_pred) *100 #Probabilidad de la variable
print("La probabilidad de la variable es:\n", pvN)

El total de datos es:
 11
La cantidad de datos en la variable Admitted es:
 No     9
Yes    2
Name: count, dtype: int64
La probabilidad de la variable es:
 No     81.818182
Yes    18.181818
Name: count, dtype: float64


4. Se realiza la evaluación del comportamiento del modelo

In [ ]:
ydp=mar.predict(XD) #Pronóstico sobre la base de datos original
cm=confusion_matrix(yd,ydp)
print("La matriz de confusión es: ",cm)

VN=cm[0,0] #Verdadero Negativo
FP=cm[0,1] #Falso Positivo
FN=cm[1,0] #Falso Negativo
VP=cm[1,1] #Verdadero Positivo

#Metricas de Desempeño
Exactitud=(VP+VN)/(VP+VN+FP+FN)
print("La Exactitud  es:\n", Exactitud)

TE= (FN+FP)/(VP+VN+FP+FN)
print("La tasa de error  es:\n", TE)

Sensibilidad=VP/(VP+FN)
print("La sensibilidad del modelo es:\n", Sensibilidad)

Especificidad=VN/(VN+FP)
print("La especificidad del modelo es:\n", Especificidad)

Precision=VP/(VP+FP)
print("La precisión del modelo es:\n", Precision)

PrNeg=VN/(VN+FP)
print("La probabilidad de Negación del modelo es:\n", PrNeg)

La matriz de confusión es:  [[10719  1297]
 [ 1985  3338]]
La Exactitud  es:
 0.8107157275506085
La tasa de error  es:
 0.18928427244939156
La sensibilidad del modelo es:
 0.627089986849521
La especificidad del modelo es:
 0.892060585885486
La precisión del modelo es:
 0.7201725997842503
La probabilidad de Negación del modelo es:
 0.892060585885486


**Análisis de Resultados:** La matriz de confusión muestra una predominancia por los valores verdadero negativos (10719) y los valores verdaderos positivos (3338). Lo anterior hizo que el modelo alcanzará valores por encima del 70% promedio para los indices de exactitud, precision, especificidad, lo que muestra claramente el buen funcionamiento del modelo frente al pronóstico de admisión. Asimismo, también se puede resaltar un valor relativamente alto en la sensibilidad, el cual se presento en aproximadamente 63%.
Adémas, podemos concluir que el modelo es de gran confiabilidad debido a que la tasa de error es bastante baja, esta se encuntra por debajo del 20%.

Por otro lado, con respecto al pronóstico de los nuevos aplicantes, se presentaron 11 aplicantes, de los cuales 2 fueron aceptados a sus carerras (18.18%) y 9 fueron rechazados por admisiones (81.81%), lo cual destaca que el numero de rechazados esta por encima de los aceptados.
Esto lo podemos evidenciar en la probalidad de negación del modelo ya que este se encuentra por encima del 85%

Para concluir, no encontramos la prescencia de nodos puros para el árbol asociado con la estructura del modelo. Sin embargo, si hubo prescencia de nodos muy cercanos a ser puros, ya que identificamos ginis de 0.042 , 0.09